File: 15-3Sum /main.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let res = [];

    nums.sort((a, b) => a - b);

    for (let index = 0; index < nums.length; index++) {
        let l = index + 1;
        let r = nums.length - 1;

        if (index > 0 && nums[index] === nums[index - 1]) continue

        while (l < r) {
            const sum = nums[index] + nums[l] + nums[r];

            if (sum < 0) {
                l++
            } else if (sum > 0) {
                r--;
            } else {
                res.push([nums[index], nums[l], nums[r]]);
                l++

                while(nums[l - 1] === nums[l] && l < r) {
                    l++
                }
            }
        }
    }



    return res

};

File: 15-3Sum /main.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	var result [][]int

	sort.Ints(nums)

	for i, num := range nums {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		l, r := i+1, len(nums)-1

		for l < r {
			res := num + nums[l] + nums[r]

			if res > 0 {
				r--
			} else if res < 0 {
				l++
			} else {
				result = append(result, []int{num, nums[l], nums[r]})
				l++

				for nums[l] == nums[l-1] && l < r {
					l++
				}
			}
		}
	}

	return result
}
